code/week6/main.cpp

```diff
@@ -1,93 +1,104 @@
-#include <opencv2/opencv.hpp>
-#include <string>
+#include <iostream>
+#include <regex>
+#include <random>
+#include "Board.h"
 
-int main() {
-    // Mat http://docs.opencv.org/3.3.0/d3/d63/classcv_1_1Mat.html
-    // this creates an image header. The array is not allocated yet
-    cv::Mat my_image;
-
-    /* This creates a matrix
-     * cv::Mat(rows, cols, type, default color)
-     * type: CV_[The number of bits per item][Signed or Unsigned][Type Prefix]C[The channel number]
-     * this type CV_8UC3
-     * 8 is the number of bits per item
-     * U means that it is unsigned
-     * C3 means that the color channel has three entires (BGR)
-    */
-    my_image = cv::Mat(2, 2, CV_8UC3, cv::Scalar(0,0,255));
-
-    // a single pixel is made up of 3 8 bit values
-    // Blue, Green, Red
-    std::cout << "my_image = \n" << my_image << std::endl;
-
-    // iterate over an image the safe way
-    for(auto it = my_image.begin<cv::Vec3b>(); it != my_image.end<cv::Vec3b>(); it++) {
-        std::cout << "Blue = " << std::to_string((*it)[0]) << std::endl;
-        std::cout << "Green = " << std::to_string((*it)[1]) << std::endl;
-        std::cout << "Red = " << std::to_string((*it)[2]) << std::endl;
-    }
-    std::cout << "\n";
+using namespace std;
 
-    // faster way to iterate over an image
-    int channels = my_image.channels();
-    std::cout << "my_image has " << channels << " channels" << std::endl;
-    int nRows = my_image.rows;
-    std::cout << "my_image has " << nRows << " rows" << std::endl;
-    std::cout << "my_image has " << my_image.cols << " cols" << std::endl;
-    // here we multiply by the number of channels since we will be dereferencing it with uchar
-    // and each channel is a uchar
-    int nCols = my_image.cols * channels;
+void generateComputerMove(const Board &board, unsigned int &r_out, unsigned int &c_out) {
+  static random_device rand_dev;
+  static default_random_engine engine(rand_dev());
+  static uniform_int_distribution<unsigned int> uniform_dist(0,2);
+  do {
+    r_out = uniform_dist(engine);
+    c_out = uniform_dist(engine);
+  } while(board.getMarker(r_out, c_out) != Marker::Empty);
```

Comment:
Won't this infinite loop if the board is full?

Reply:
The main method checks isOver before calling this method, so in practice board will not be full.
Were this a more generic code base, yes, we'd want to check that here.